workshops/filters.py

```diff
@@ -207,7 +222,218 @@ def get_order_by(self, order_value):
+
+
+def filter_training_request(queryset, flag):
+    if flag is None:
+        return queryset
+    elif flag:
+        return queryset.filter(trainingrequest__isnull=False).distinct()
+    else:
+        return queryset.filter(trainingrequest__isnull=True)
+
+
+def filter_training(queryset, training):
+    if training is None:
+        return queryset
+    else:
+        return queryset.filter(task__role__name='learner',
+                               task__event=training).distinct()
+
+
+class TraineeFilter(AMYFilterSet):
+    search = django_filters.CharFilter(
+        action=filter_trainee_name,
+        label='Name or Email')
+
+    all_persons = django_filters.BooleanFilter(
+        label='Include all people, not only trainees',
+        action=filter_all_persons,
+        widget=widgets.CheckboxInput)
+
+    homework = django_filters.BooleanFilter(
+        label='Only trainees with unevaluated homework',
+        widget=widgets.CheckboxInput,
+        action=filter_unevaluated_homework,
+    )
+
+    training_request = django_filters.BooleanFilter(
+        label='Is training request present?',
+        action=filter_training_request,
+    )
+
+    is_swc_instructor = django_filters.BooleanFilter(
+        label='Is SWC Instructor?',
+        name='is_swc_instructor',
+    )
+
+    is_dc_instructor = django_filters.BooleanFilter(
+        label='Is DC Instructor?',
+        name='is_dc_instructor',
+    )
+
+    training = django_filters.ModelChoiceFilter(
+        queryset=Event.objects.ttt(),
+        action=filter_training,
+    )
+
+    class Meta:
+        model = Person
+        fields = [
+            'search',
+            'all_persons',
+            'homework',
+            'is_swc_instructor',
+            'is_dc_instructor',
+            'training',
+        ]
+        order_by = ["-last_login", "lastname", "-lastname", "firstname", "-firstname",
+                    "email", "-email"]
+
+    def get_order_by(self, order_value):
+        if order_value == 'firstname':
+            return ['personal', 'middle', 'family']
+        elif order_value == '-firstname':
+            return ['-personal', '-middle', '-family']
+        elif order_value == 'lastname':
+            return ['family', 'middle', 'personal']
+        elif order_value == '-lastname':
+            return ['-family', '-middle', '-personal']
+        else:
+            return super().get_order_by(order_value)
+
+
+def filter_matched(queryset, choice):
+    if choice == '':
+        return queryset
+    elif choice == 'u':  # unmatched
+        return queryset.filter(person=None)
+    elif choice == 'p':  # matched trainee, unmatched training
+        return queryset.filter(person__isnull=False)\
+                       .exclude(person__task__role__name='learner',
+                                person__task__event__tags__name='TTT')\
+                       .distinct()
+    else:
+        assert choice == 't'  # matched trainee and training
```

Comment:
No. If you provide `choices` to the field, django-filter should say "this is not a valid choice" on its own when you use a wrong choice. At least it did when I tested similar scenario a few months ago.